### PR TITLE
feat: Set CORS headers

### LINKED ISSUE #4984: Cross Origin Requests
# Summary

Allow CORS by adding a way to set the "Access-Control-Allow-Origin" headers

What change needs making?

Something like 

```
argo-server --access-control-allow-origin="http://localhost:8080"
```

# Use Cases

When would you use this?

For making requests from my web application to Argo under the origin of `http://localhost:8080` for example.
I can issue a PR for this since this is just setting an HTTP header.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by joyciep**
Set `Access-Control-Allow-Origin` header by passing the allowed origins via `--access-control-allow-origin` flag

```
argo-server --access-control-allow-origin="<origin>"
```

Fixes #4984 

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: server/apiserver/argoserver.go [hunk 1]
```diff
@@ -300,17 +303,34 @@ func (as *argoServer) newHTTPServer(ctx context.Context, port int, artifactServe
...
 	mux.HandleFunc("/oauth2/redirect", as.oAuth2Service.HandleRedirect)
 	mux.HandleFunc("/oauth2/callback", as.oAuth2Service.HandleCallback)
 	mux.Handle("/metrics", promhttp.Handler())
 	// we only enable HTST if we are secure mode, otherwise you would never be able access the UI
-	mux.HandleFunc("/", static.NewFilesServer(as.baseHRef, as.tlsConfig != nil && as.hsts, as.xframeOptions).ServerFiles)
+	mux.HandleFunc("/", configureCORS(static.NewFilesServer(as.baseHRef, as.tlsConfig != nil && as.hsts, as.xframeOptions).ServerFiles, as.accessControlAllowOrigin))
 	return &httpServer
 }
 
+// configureCORS is a middleware for setting CORS headers
```
Comment (alexec) [inline server/apiserver/argoserver.go:317]: Looks good. Any reason to not insert this next to the x-frames-options code?

Comment (joyciep) [reply server/apiserver/argoserver.go:317]: I just want this middleware close to the callers. Should I move it?

Comment (alexec) [reply server/apiserver/argoserver.go:317]: Look for where we already add headers suck as X-Frame-Options

Comment (joyciep) [reply server/apiserver/argoserver.go:317]: Okay found it. Should I not use the middleware and just call `w.Header().Set()` in `static.go`?

Comment (joyciep) [reply server/apiserver/argoserver.go]: Moved it near the x-frame-options code, kindly check.

**Review by alexec — APPROVED** [review]
LGTM

File: server/apiserver/argoserver.go (hunk 1 above)
Comment (uhlajs) [reply server/apiserver/argoserver.go]: @joyciep After this change, are the CORS configured also for `/api/` URI? If I am reading it right, it is only used for `static.NewFilesServer`, isn't it?

Comment (KimPaow) [reply server/apiserver/argoserver.go]: @joyciep @uhlajs @alexec

Sorry for the big mention such a long time after this was closed. 

I'm trying to use the HTTP api of an instance of argo-server running in Kubernetes from localhost. I've passed the `--access-control-allow-origin=*` arg but am still running into CORS issues on my client app.

Seeing as this was merged before the last question was answered; I'm wondering, does the `/api/` routes also use the cors arg?

I'm not literate enough in Go to understand myself what's going on.

**Pull request merged**